### PR TITLE
feat(frontend): utxos-fee context

### PR DESCRIPTION
**Pull request by DenysKarmazynDFINITY**
# Motivation

The idea of this PR is pretty similar to what we already have for `bitcoin-fee.store` - we need a context for storing and fetching utxos fee on evert `amount` change. This store will be also useful for BTC send flow.

### CODE REVIEW

File: src/frontend/src/btc/stores/utxos-fee.store.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export type UtxosFeeStoreData = Option<{
+	utxosFee?: UtxosFee;
+}>;
+
+export interface UtxosFeeStore extends Readable<UtxosFeeStoreData> {
+	setUtxosFee: (data: UtxosFeeStoreData) => void;
+}
+
+export const initUtxosFeeStore = (): UtxosFeeStore => {
```
Comment (peterpeterparker) [inline src/frontend/src/btc/stores/utxos-fee.store.ts:13]: Can you provide dedicated tests for the store?

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		}
+
+		store.setUtxosFee({
+			utxosFee
+		});
+	};
+
+	const debounceEstimateFee = debounce(loadEstimatedFee);
+
+	$: amount, networkId, (() => debounceEstimateFee())();
```
Comment (peterpeterparker) [inline src/frontend/src/btc/components/fee/UtxosFeeContext.svelte:54]: You might be able to simplify

```
$: amount, networkId, debounceEstimateFee()
```

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	import type { NetworkId } from '$lib/types/network';
+	import { isNetworkIdBitcoin, mapNetworkIdToBitcoinNetwork } from '$lib/utils/network.utils';
+
+	export let amount: number | undefined = undefined;
+	export let networkId: NetworkId | undefined = undefined;
+
+	const { store } = getContext<UtxosFeeContext>(UTXOS_FEE_CONTEXT_KEY);
+
+	const loadEstimatedFee = async () => {
+		if (isNullish($authIdentity)) {
```
Comment (peterpeterparker) [inline src/frontend/src/btc/components/fee/UtxosFeeContext.svelte:17]: Should we have a test for this edge case?

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			return;
+		}
+
+		if (
+			isNullish(networkId) ||
+			!isNetworkIdBitcoin(networkId) ||
+			isNullish(amount) ||
+			amount === 0
+		) {
+			store.setUtxosFee(null);
```
Comment (peterpeterparker) [inline src/frontend/src/btc/components/fee/UtxosFeeContext.svelte:28]: Is there a test for this use case?

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+			return;
+		}
+
+		const network = mapNetworkIdToBitcoinNetwork(networkId);
+
+		const utxosFee = nonNullish(network)
+			? await selectUtxosFeeApi({
```
Comment (peterpeterparker) [inline src/frontend/src/btc/components/fee/UtxosFeeContext.svelte:35]: ditto

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+					amount,
+					network,
+					identity: $authIdentity
+				})
+			: undefined;
+
+		if (isNullish(utxosFee)) {
+			store.setUtxosFee(null);
```
Comment (peterpeterparker) [inline src/frontend/src/btc/components/fee/UtxosFeeContext.svelte:43]: Do you want to add a function `reset` to the store instead of accepting `null`?

File: src/frontend/src/btc/stores/utxos-fee.store.ts (hunk 1 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/stores/utxos-fee.store.ts:14]: I created initially a few tests there, now added additional test for `reset`. Please let me know if you have some other cases in mind.

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte (hunk 2 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/fee/UtxosFeeContext.svelte]: True, done

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte (hunk 3 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/fee/UtxosFeeContext.svelte:17]: Added.

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte (hunk 4 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/fee/UtxosFeeContext.svelte]: Now yes :)

File: src/frontend/src/btc/components/fee/UtxosFeeContext.svelte (hunk 6 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/fee/UtxosFeeContext.svelte]: yes, sounds good, added

**Comment by DenysKarmazynDFINITY** [comment]
@peterpeterparker please take another look

File: src/frontend/src/btc/stores/utxos-fee.store.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	const { subscribe, set } = writable<UtxosFeeStoreData>(undefined);
+
+	return {
+		subscribe,
+
+		reset() {
+			set(undefined);
```
Comment (peterpeterparker) [inline src/frontend/src/btc/stores/utxos-fee.store.ts:21]: I think we use `set(null)`. Can you double check?

File: src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	const mockAuthStore = (value: Identity | null) =>
+		vi.spyOn(authStore, 'authIdentity', 'get').mockImplementation(() => readable(value));
+
+	const props = {
+		amount,
+		networkId
+	};
+
+	beforeEach(() => {
+		mockPage.reset();
```
Comment (peterpeterparker) [inline src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts:34]: reset the utxos store as well on before?

File: src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+		});
+	});
+
+	it('should not call selectUtxosFee if no networkId provided', async () => {
+		const store = initUtxosFeeStore();
+		const resetSpy = vi.spyOn(store, 'reset');
+		const selectUtxosFeeSpy = mockBtcSendApi();
+		const { networkId: _, ...newProps } = props;
+
+		mockAuthStore(mockIdentity);
```
Comment (peterpeterparker) [inline src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts:83]: To avoid duplicating `mockAuthStore(mockIdentity);` you can have two distinct describe with a before each

```
describe("not signed-in", () =>
    beforeEach
       mockAuthStore(undefined);

describe("signed-in", () =>
   beforeEach
      mockAuthStore(mockIdentity);
```

Just sharing if interesting, not a call to action.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx!

If you can have a look to the new comments before merge, that's super.

File: src/frontend/src/btc/stores/utxos-fee.store.ts (hunk 7 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/stores/utxos-fee.store.ts:21]: True, changed to `null`

File: src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts (hunk 8 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts:34]: yep, done

File: src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts (hunk 9 above)
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/tests/btc/components/fee/UtxosFeeContext.spec.ts:83]: I kept the initial structure but shortened it a bit by using `mockIdentity` as default value and passing `null` only in the related test.
